trie/binary.go

```diff
@@ -0,0 +1,154 @@
+	value []byte
+	db    ethdb.Database
+}
+
+func NewBinary(db ethdb.Database) (*BinaryTrie, error) {
+	if db == nil {
+		return nil, fmt.Errorf("trie.NewBinary called without a database")
+	}
+
+	return &BinaryTrie{db: db}, nil
+}
+
+func (t *BinaryTrie) Get(key []byte) []byte {
+	res, err := t.TryGet(key)
+	if err != nil {
+		log.Error(fmt.Sprintf("Unhandled trie error: %v", err))
+	}
+	return res
+}
+
+func (t *BinaryTrie) TryGet(key []byte) ([]byte, error) {
+	value, err := t.tryGet(key, 0)
+	return value, err
+}
+
+func (t *BinaryTrie) tryGet(key []byte, depth int) ([]byte, error) {
+	by := key[depth/8]
+	bi := (by >> uint(depth%8)) & 1
+	if bi == 0 {
+		if t.left == nil {
+			if depth < len(key)*8-1 || t.value == nil {
+				return nil, fmt.Errorf("could not find key %s in trie", common.ToHex(key))
+			}
+			return t.value, nil
+		}
+		return t.left.tryGet(key, depth+1)
+	} else {
+		if t.right == nil {
+			if depth < len(key)*8-1 || t.value == nil {
+				return nil, fmt.Errorf("could not find key %s in trie", common.ToHex(key))
+			}
+			return t.value, nil
+		}
+		return t.left.tryGet(key, depth+1)
+	}
+}
+
+func (t *BinaryTrie) Update(key, value []byte) {
+	if err := t.TryUpdate(key, value); err != nil {
+		log.Error(fmt.Sprintf("Unhandled trie error: %v", err))
+	}
+}
+
+func (t *BinaryTrie) TryUpdate(key, value []byte) error {
+	// TODO check key depth
+	err := t.insert(0, key, value)
+	return err
+}
+
+func (t *BinaryTrie) insert(depth int, key, value []byte) error {
+	// TODO hash intermediate nodes
+	by := key[depth/8]
+	bi := (by >> uint(depth%8)) & 1
+	if bi == 0 {
+		if t.left == nil {
+			if depth == len(key)*8-2 {
+				t.left = &BinaryTrie{nil, nil, value, t.db}
+				return nil
+			} else {
+				t.left = &BinaryTrie{nil, nil, nil, t.db}
+			}
+		}
+		return t.left.insert(depth+1, key, value)
+	} else {
+		if t.right == nil {
+			if depth == len(key)*8-2 {
+				t.right = &BinaryTrie{nil, nil, value, t.db}
+				return nil
+			} else {
+				t.right = &BinaryTrie{nil, nil, nil, t.db}
+			}
+		}
+		return t.right.insert(depth+1, key, value)
+	}
+}
+
+func (t *BinaryTrie) Commit() ([]byte, error) {
+	var payload [3][]byte
+	var err error
+	if t.left != nil {
+		payload[0], err = t.left.Commit()
+		if err != nil {
+			return nil, err
+		}
+	}
+	if t.right != nil {
+		payload[1], err = t.right.Commit()
+		if err != nil {
+			return nil, err
+		}
+	}
+	hasher := sha3.NewLegacyKeccak256()
+	if t.value != nil {
+		hasher.Write(t.value)
+		hasher.Sum(payload[2][:])
+	}
+
+	hasher.Reset()
+	hasher.Write(payload[0])
+	hasher.Write(payload[1])
+	hasher.Write(payload[2])
+	h := hasher.Sum(nil)
+	data := make([]byte, len(payload[0])+len(payload[1])+len(payload[2])+3)
+	data[0] = byte(len(payload[0]))
+	copy(data[1:], payload[0])
+	data[len(payload[0])+1] = byte(len(payload[1]))
+	copy(data[2+len(payload[0]):], payload[1])
+	data[len(payload[0])+len(payload[1])+2] = byte(len(payload[2]))
+	copy(data[2+len(payload[0])+len(payload[1]):], payload[2])
```

Comment:
Where is this format defined?

Reply:
It's a pure flight of fancy for the moment, I wanted to see what it would take to convert the trie

Reply:
I changed the format to be closer to what the hex prefix does.